lib/active_merchant/billing/gateways/flo2cash_rest.rb

```diff
@@ -0,0 +1,595 @@
+      self.display_name = 'Flo2Cash'
+
+      # Errors Codes from flo2Cash
+      #
+      #   05  Declined – Bank declined
+      #   10  Declined – Bank error
+      #   51  Declined – Insufficient funds
+      #   54  Declined – Expired card
+      #   68  Failed – No reply from bank
+      #
+      STANDARD_ERROR_CODE_MAPPING = {
+        '05' => STANDARD_ERROR_CODE[:processing_error],
+        '10' => STANDARD_ERROR_CODE[:processing_error],
+        '51' => STANDARD_ERROR_CODE[:card_declined],
+        '54' => STANDARD_ERROR_CODE[:expired_card],
+        '68' => STANDARD_ERROR_CODE[:processing_error],
+      }
+
+      BRAND_MAP = {
+        "visa" => "VISA",
+        "master" => "MC",
+        "american_express" => "AMEX",
+        "diners_club" => "DINERS"
+      }
+
+      CLIENT_ERROR_REGEX = /^4\d\d/
+
+      # The following are the response statuses from flo2cash to each endpoint
+      #
+      # If a transaction results in a decline due to reasons like card expired
+      # or insufficient funds, the transaction will be marked as Failed
+      #
+      # PAYMENT_STATUSES = %w(successful declined blocked failed unknown)
+      # CARD_PLAN_STATUSES = %w(active suspended ended cancelled)
+      # DIRECT_DEBIT_PLAN_STATUSES = %w(pending-approval active suspended ended cancelled)
+      # DIRECT_DEBIT_STATUSES = %w(scheduled processing successful dishonoured failed)
+
+      SUCCESS_STATUSES = %w(successful scheduled processing active cancelled pending-approval)
+
+      ACTION_URL_MAP = {
+        store: 'cardTokens',
+        purchase: '/payments',
+        refund: "/payments/%{payment_id}/refunds",
+        create_card_plan: '/cardplans',
+        create_direct_debit_plan: "/directdebitplans",
+        update_card_plan: "/cardplans/%{plan_id}/status",
+        update_direct_debit_plan: "/directdebitplans/%{plan_id}/status",
+        retrieve_card_plan: "/cardplans/%{plan_id}",
+        retrieve_card_payment: "/payments/%{payment_id}",
+        retrieve_direct_debit_plan: "/directdebitplans/%{plan_id}",
+      }
+
+      # Creates a new Flo2cachRest gateway
+      #
+      # The gateway requires a +merchant_id+ and a +api_key+ key be passed in
+      # the +options+ hash.
+      #
+      # ==== Options
+      #
+      # * <tt>:merchant_id</tt> -- The Flo2Cash Merchant ID (REQUIRED)
+      # * <tt>:api_key</tt> -- The Flo2Cash API key (REQUIRED)
+      #
+      def initialize(options = {})
+        requires!(options, :merchant_id, :api_key)
+        super
+      end
+
+      # Process Payment
+      #
+      def purchase(money, payment_method, options={})
+        post = {}
+        add_type(post, 'purchase')
+        add_channel(post, options)
+        add_invoice(post, money, options)
+        add_customer_resume(post, options)
+        add_references(post, options)
+        add_device_info(post, options)
+        add_geolocation(post, options)
+
+        token = retrieve_token(payment_method, options)
```

Comment:
I was confused by this method. If I understood it correctly, you are testing if the `payment_method` passed in to the method is the token then you will use that, otherwise you will enforce a tokenisation? Is it a requirement for a `purchase` to _always_ use a tokenised card number?

Reply:
Yes, a `payment` (`purchase`) only support two way to perform... one with token and other with iframe.

```

Reply:
Ok, so instead of having a private method, why not put the logic in here?